[mssql] Fix connection corruption after executing sql

nyalldawson opened the pull request:
Executing any SQL on a SQL Server connection would result in the connection being dropped while it was still in use, leaving the provider in a interdeterminate state which would often refuse to run futher queries.

The QSqlQuery class does NOT have a virtual destructor, so we need to explicitly use the QgsMssqlQuery class for the results iterator so that the QgsMssqlQuery destructor gets called and the shared reference to the QgsMssqlDatabase instance gets released correctly.

nyalldawson committed:
[mssql] Fix connection corruption after executing sql

Executing any SQL on a SQL Server connection would result in the
connection being dropped while it was still in use, leaving
the provider in a interdeterminate state which would often
refuse to run futher queries.

The QSqlQuery class does NOT have a virtual destructor, so we
need to explicitly use the QgsMssqlQuery class for the results
iterator so that the QgsMssqlQuery destructor gets called and
the shared reference to the QgsMssqlDatabase instance gets
released correctly.

Gabriel-Diosan commented:
Tested this patchset. The driver not loaded error still happens after deleting a table from the MS SQL database. The SQL log doesn't show any error messages. Once the driver not loaded error is displayed, executing SQL commands works at times but after a while you get a Connection is busy with results for another hstmt QODBC3.
![QGIS_Import_Failure_18_March_2025](https://github.com/user-attachments/assets/0759e237-2603-4fca-81bc-eccb5788e936)
![QGIS_Connection_Busy](https://github.com/user-attachments/assets/52b4adc7-e0be-4ecf-adba-07ebfd9b5ea3)

@nyalldawson I will send you a copy of the log shortly.

nyalldawson committed:
[cleanup] Avoid redundant container lookups

nyalldawson committed:
[mssql] Use method which takes an existing connection instead of creating new connection

We already have a connection here:

nyalldawson committed:
[mssql] Optimise geom column type resolution

Don't create a new connection per table, when we can
just use one connection for all

nyalldawson committed:
More thread safety for QgsVectorLayerExporterTask

nyalldawson committed:
Fix leak after creating table on sql server

nyalldawson committed:
Fix crash when db import dialog closes

nyalldawson commented:
@Gabriel-Diosan can you test this new version?

Gabriel-Diosan commented:
@nyalldawson I retested with the new build and I see the following. Listing tables takes quite some time and adding a new vector layer to the SQL database causes QGIS to lockup. A table is created but it is empty with no data. Additionally, becuase the schema drop down does not work in the build, all layers are imported into dbo. Not sure whether there are some missing patches to do with the schema dropdown that would help the situation?

nyalldawson commented:
@Gabriel-Diosan 
Hmm, are all those issues regressions in this build alone, or do you see them in master builds too? There's nothing here which should regress the speed of table listings or would explain the empty tables. 

(It's quite possible those are issues in the mingw/qt 6 builds alone, unrelated to this work)

Gabriel-Diosan commented:
@nyalldawson So it turns out the issue with tables loading may be to do with the test database having limited resources as the currently nightly and QGIS 3.40 both have the same issue at the moment. I tried this build with our production database and get the error below when trying to import a layer.
![Geometry_Columns_Error](https://github.com/user-attachments/assets/9cb75997-abf7-40ca-8f23-b12451519960)

nyalldawson commented:
@Gabriel-Diosan 

Ah, thanks! The geometry columns table issue sounds familiar, that's https://github.com/qgis/QGIS/issues/57614. I'll take a look what's involved in fixing that

nyalldawson commented:
@Gabriel-Diosan fix in https://github.com/qgis/QGIS/pull/61101

nyalldawson committed:
Remove unused dead code

Probably a remnant of copy/paste from the original implementation
from the postgres provider

nyalldawson committed:
[mssql] Fix native integer types

bigint should map to LongLong variant type

Add actual type names to display strings

nyalldawson committed:
[mssql] Fix fields get unaligned when importing data
